kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,717 @@
+            env=env,
+        )
+        pipe.wait()
+        if pipe.returncode:
+            raise RuntimeError("Non zero exit status from ansible (%s)" % pipe.returncode)
+
+    def _get_kubeconfig(self, ip):
+        cat_kubeconf = "sudo cat /etc/kubernetes/admin.conf"
+        host = "@".join((self.ssh_username, ip))
+        args = ("ssh", host) + self.ssh_common_args + (cat_kubeconf,)
+        kubeconfig = yaml.safe_load(subprocess.check_output(args))
+        kubeconfig["clusters"][0]["cluster"]["server"] = "https://%s:6443" % ip
+        return kubeconfig
+
+    def get_ssh_key(self):
+        """Generate ssh keypair if not exist.
+
+        Return public key as string.
+        """
+        os.makedirs(config.KS_FILES_PATH, exist_ok=True)
+        ssh_key_path = os.path.join(config.KS_FILES_PATH, "ssh_key")
+        if not os.path.exists(ssh_key_path):
+            cmd = [config.KS_SSH_KEYGEN_CMD, "-P", "", "-f", ssh_key_path]
+            subprocess.check_call(cmd)
+        with open(ssh_key_path + ".pub", "r") as key_file:
+            return key_file.read()
+
+
+class OpenStack:
+    """Openstack client wrapper."""
+
+    def __init__(self, stack_name, *, os_kwargs, cluster, extra_ssh_key):
+        self.c = OpenStack.get_connection(os_kwargs)
+        self.c.authorize()
+        self.cluster = cluster
+        self.extra_ssh_key = extra_ssh_key
+        self.stack_name = stack_name
+        self.os_kwargs = os_kwargs
+
+    def provision(self):
+        master_count = self.cluster.metadata["master_count"]
+        slave_count = self.cluster.metadata["slave_count"]
+        dns = self.cluster.metadata["dns_nameservers"].split(",")
+        ext_net = self.c.get_network(self.cluster.metadata["floating_network"])
+        if ext_net is None:
+            raise Exception("External network %s not found" % self.cluster.metadata["floating_network"])
+        image = self.c.get_image(self.cluster.metadata["image_name"])
+        if image is None:
+            raise Exception("Image %s not found" % self.cluster.metadata["image_name"])
+        flavor = self.c.get_flavor(self.cluster.metadata["flavor"])
+        if flavor is None:
+            raise Exception("Flavor %s not found" % self.cluster.metadata["flavor"])
+        resources = {
+            "masters": [],
+            "slaves": [],
+        }
+        network = self.c.create_network(self.stack_name)
+        subnet = self.c.create_subnet(network, cidr="10.1.0.0/16",
+                                      subnet_name=self.stack_name,
+                                      dns_nameservers=dns)
+        router = self.c.create_router(name=self.stack_name,
+                                      ext_gateway_net_id=ext_net.id)
+        self.c.add_router_interface(router, subnet["id"])
+        resources["router_id"] = router["id"]
+        resources["network_id"] = network["id"]
+        resources["subnet_id"] = subnet["id"]
+        # TODO: security group
```

Comment:
should we add it later? could you please describe what did you mean here

Reply:
There is a problem with security groups on some clouds. I'll better remove this comment at all.